utils/pattern/src/replacement.rs

```diff
@@ -0,0 +1,203 @@
+/// ]);
+/// replacements.insert(5, vec![
+///     Element::TokenFive
+/// ]);
+///
+/// let mut parser = Parser::new("{5}, {0}");
+/// let mut interpolator = Interpolator::new(parser, replacements);
+///
+///
+/// assert_eq!(Ok(Some(Element::TokenFive)), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Literal(", ".to_string()))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::TokenZero)), interpolator.try_next());
+/// assert_eq!(Ok(None), interpolator.try_next());
+/// ```
+///
+/// # Advanced Replacement Provider
+///
+/// Another scenario in which a consumer may want to implement their own `ReplacementProvider`
+/// is when they can avoid allocation and storing of the elements, and instead can stream the
+/// replacements or compute them on the fly.
+///
+/// To illustrate such an example, we'll create a custom iterator which computes digits from `1`
+/// to some `max` value and return such iterator from a custom replacement provider.
+///
+/// ## Examples
+/// ```
+/// use icu_pattern::{Parser, Interpolator, ReplacementProvider};
+///
+/// #[derive(Debug, PartialEq)]
+/// enum Element {
+///     Digit(usize),
+///     Literal(String)
+/// }
+///
+/// impl From<&str> for Element {
+///     fn from(input: &str) -> Self {
+///         Self::Literal(input.to_string())
+///     }
+/// }
+///
+/// struct MyReplacementProvider;
+///
+/// struct MyIterator {
+///     idx: usize,
+///     max: usize,
+/// }
+///
+/// impl MyIterator {
+///     pub fn new(max: usize) -> Self {
+///         Self {
+///             idx: 0,
+///             max
+///         }
+///     }
+/// }
+///
+/// impl Iterator for MyIterator {
+///     type Item = Element;
+///
+///     fn next(&mut self) -> Option<Self::Item> {
+///         self.idx += 1;
+///         if self.idx <= self.max {
+///             Some(Element::Digit(self.idx))
+///         } else {
+///             None
+///         }
+///     }
+/// }
+///
+///
+/// impl ReplacementProvider<usize, Element> for MyReplacementProvider {
+///     type Iter = MyIterator;
+///
+///     fn take_replacement(&mut self, key: &usize) -> Option<Self::Iter> {
+///         Some(MyIterator { idx: 0, max: *key })
+///     }
+/// }
+///
+/// let mut parser = Parser::new("{4}, {2}");
+/// let mut interpolator = Interpolator::new(parser, MyReplacementProvider);
+///
+/// assert_eq!(Ok(Some(Element::Digit(1))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Digit(2))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Digit(3))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Digit(4))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Literal(", ".to_string()))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Digit(1))), interpolator.try_next());
+/// assert_eq!(Ok(Some(Element::Digit(2))), interpolator.try_next());
+/// assert_eq!(Ok(None), interpolator.try_next());
+/// ```
+pub trait ReplacementProvider<K, E> {
+    type Iter: Iterator<Item = E>;
```

Comment:
Issue: I do not understand the decision behind making ReplacementProvider return an iterator for each placeholder.  I would have expected it to provide a scalar instead (either a string or something convertible to a string).  It seems like an iterator here is overly complex.  If you have a good reason for using an iterator, I would at least suggest that the default ReplacementProvider impls work on `Vec<E>` and `HashMap<K, E>`, and use [std::iter::once](https://doc.rust-lang.org/std/iter/fn.once.html) to convert the scalars into iterators.

Reply:
Returning an iterator is necessary for the logic to work. `Interpolator` servers like `zipper` that takes placeholder patter iterator, and replacement iterators and zips them together.

> If you have a good reason for using an iterator

The reason is to allow for zero-allocation zipping. The `Interpolator` doesn't allocate, the `Parser` doesn't allocate, and the replacement parsers may also not allocate, just stream through something (or compute it) and stream tokens back.

> I would at least suggest that the default ReplacementProvider impls work on Vec<E> and HashMap<K, E>

I can add that! Can you give me an example of a scenario where the replacement is one-item?

Reply:
> just stream through something (or compute it) and stream tokens back.

The `Interpolator` can stream through a single `E` token; there's no reason it needs to stream through an iterator of `E`s.

> Can you give me an example of a scenario where the replacement is one-item?

Like... the most common case?  "In {0} days" take a number token.  "{0}, {1}, and {2}" takes three scalar tokens to form the list.

I'm stretching my imagination to think of cases when you'd want an iterator.  I guess if you have the glue pattern "{0} 'at' {1}", you could attach it to a sub-pattern for the date and the time which may be composed themselves of multiple tokens.

Reply:
Right, what your imagination is stretched by is exactly the use case I was working on :)

Thanks, I'll think more of the `In {0} days` case. I have not planned this iterator to be used in such a manner because in my experience NF and DTF both have their own parsers with their own symbols and the placeholder pattern is used more a combination pattern.
But it seems like what you're saying is that it is common that we will use this parser as the "only" parser, rather than a combinator and in those cases single-items are needed.

I think I can extract a more general interpolator for which a combining one is a special case of! I'll work on it.

Reply:
Added `ReplacementProvider` for `Vec<E>` - it messes up signatures a bit since now we have to explicitly distinguish between `Vec<Vec<E>>` and `Vec<E>` but I think it's worth it and it's an internal code anyway, so I'm sacrificing a bit of ergonomics of Rust guessing your generics for the general usability and performance of the code.

I'll continue investigating ways to make the API nicer for the simple case, and I'm still baffled that I spent a month working on this code not considering the "simple" case at all! :)